sparta/src/Scoreboard.cpp

```diff
@@ -0,0 +1,417 @@
+        // large as the number of rows in the matrix (minus 1 for the
+        // header)
+        const auto num_producers = params->latency_matrix.getValue().size() - 1;
+        forwarding_latencies_.resize(num_producers);
+        unit_id_to_scoreboard_views_.resize(num_producers);
+
+        // Skip the first row -- it's a header
+        for (uint32_t producer_row_idx = 1;
+             producer_row_idx < params->latency_matrix.getValue().size();
+             ++producer_row_idx)
+        {
+            const auto & producer_row = params->latency_matrix.getValue()[producer_row_idx];
+
+            const auto producer_name = producer_row[0];
+
+            // Remember the producer name and assign an ID to it
+            sparta_assert(unit_name_to_id_.find(producer_name) == unit_name_to_id_.end(),
+                          "Unit name already is in the latency table (column 0) twice: "
+                          << producer_name);
+            const uint32_t producer_id = unit_id_++;
+            unit_name_to_id_[producer_name] = producer_id;
+
+            // Set up the forwarding latency table column size to be
+            // as large as the number of columns in the matrix (minus 1
+            // for the producer names)
+            forwarding_latencies_[producer_id].resize(producer_row.size() - 1);
+
+            // Set up the producer -> consumer latency table
+            // Skip the first column, it's the producer name
+            for (uint32_t consumer_column_idx = 1; consumer_column_idx < producer_row.size(); ++consumer_column_idx)
+            {
+                try {
+                    forwarding_latencies_[producer_id][consumer_column_idx - 1] =
+                        std::stol(producer_row[consumer_column_idx]);
+                }
+                catch(...) {
+                    sparta_assert(false, "Error while trying to convert " <<
+                                  producer_row[consumer_column_idx] << " to an int. Row "
+                                  << producer_row_idx
+                                  << " Column "
+                                  << consumer_column_idx);
+                }
+            }
+        }
+
+    }
+
+    void Scoreboard::set(const Scoreboard::RegisterBitMask & bits)
+    {
+        // Update registered scoreboards immediately
+        for(auto & sbvs : unit_id_to_scoreboard_views_)
+        {
+            for (auto * sbv : sbvs)
+            {
+                if(nullptr != sbv)
+                {
+                    sbv->receiveScoreboardUpdate_(bits, INVALID_UNIT_ID);
+                }
+            }
+        }
+    }
+
+    void Scoreboard::set(const Scoreboard::RegisterBitMask & bits, Scoreboard::UnitID producer)
+    {
+        sparta_assert(producer < forwarding_latencies_.size(),
+                      "could not find producer ID in forwarding_latencies table");
+
+        const auto & consumers = forwarding_latencies_[producer];
```

Comment:
It'd be cool if the `forwarding_latencies_` vector *only* contained valid scoreboards worth forwarding.  That way we can get rid of the `if` statment on line 173.  Same kinda logic for line 180.

Reply:
Done. I needed an additional data structure to implement this.